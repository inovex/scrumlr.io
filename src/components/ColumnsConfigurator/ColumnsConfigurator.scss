@import "constants/style";

.columns-configurator {
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-shrink: 0;
  flex-grow: 1;

  // allow horizontal scroll
  overflow: auto hidden;
  width: 100%;
  max-width: 100%;
  height: 100%;

  @include scrollbar();
}

.columns-configurator__columns-wrapper {
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-shrink: 0;
  flex-grow: 1;
}
