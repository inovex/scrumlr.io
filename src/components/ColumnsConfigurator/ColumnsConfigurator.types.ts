export type ColumnPlacement = "first" | "center" | "last" | "all";

export type AddTemplateColumnAlignment = "left" | "right";
