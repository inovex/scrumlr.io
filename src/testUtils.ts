export type AssertTypeEqual<T, Expected> =
    T extends Expected
        ? (Expected extends T ? true : never)
        : never;