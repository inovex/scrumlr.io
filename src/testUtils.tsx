import {render, RenderOptions} from "@testing-library/react";
import {I18nextProvider} from "react-i18next";
import {FC, PropsWithChildren, ReactElement} from "react";
import {MemoryRouter} from "react-router";
import {Provider} from "react-redux";
import getTestStore from "utils/test/getTestStore";
import i18n from "./i18nTest";

export type AssertTypeEqual<T, Expected> = T extends Expected ? (Expected extends T ? true : never) : never;

const AllTheProvidersWithoutRouter: FC<PropsWithChildren> = ({children}) => (
  <I18nextProvider i18n={i18n}>
    <Provider store={getTestStore()}>{children}</Provider>
  </I18nextProvider>
);

const AllTheProviders: FC<PropsWithChildren> = ({children}) => (
  <MemoryRouter>
    <AllTheProvidersWithoutRouter>{children}</AllTheProvidersWithoutRouter>
  </MemoryRouter>
);

const customRender = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) => render(ui, {wrapper: AllTheProviders, ...options});
const customRenderWithoutRouter = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) =>
  render(ui, {wrapper: AllTheProvidersWithoutRouter, ...options});

export {customRender as render, customRenderWithoutRouter as renderWithoutRouter};
