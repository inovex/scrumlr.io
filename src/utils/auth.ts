import {Toast} from "utils/Toast";
import {API} from "api";
import i18n from "i18next";
import store from "store";
import {Actions} from "store/action";
import {SERVER_HTTP_URL} from "../config";

/**
 * Sign in anonymously.
 *
 * @param displayName Display name of the parse auth user.
 *
 * @returns Promise with user credentials on successful sign in, null otherwise.
 */
const signInAnonymously = async (displayName: string) => {
  try {
    const user = await API.signInAnonymously(displayName);
    if (user) {
      store.dispatch(Actions.signIn(user.id, user.name));
    }
    return true;
  } catch (err) {
    Toast.error({title: i18n.t("Toast.authenticationError")});
    return null;
  }
};

/**
 * Redirects to OAuth page of provider:
 *    https://accounts.google.com/o/oauth2/v2/auth/oauthchooseaccount...
 *    https://github.com/login/oauth/authorize...
 *    https://login.microsoftonline.com/common/oauth2/v2.0/authorize...
 * @param authProvider name of the OAuth Provider
 * @param originURL origin URL
 */
const signInWithAuthProvider = async (authProvider: string, originURL: string) => {
  window.location.href = `${SERVER_HTTP_URL}/login/${authProvider}?state=${encodeURIComponent(originURL)}`;
};

export const Auth = {
  signInAnonymously,
  signInWithAuthProvider,
};
