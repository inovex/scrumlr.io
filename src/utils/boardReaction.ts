import {BoardReactionType} from "types/boardReaction";

// these are custom events that can be emitted and subscribed to.
// using redux is overkill, so we decided to separate the logic, similar to toasts
export const ADD_BOARD_REACTION = "scrumlr.io/addBoardReaction";
export const REMOVE_BOARD_REACTION = "scrumlr.io/removeBoardReaction";

const add = (reaction: BoardReactionType) => {
  const addBoardReactionEvent = new CustomEvent<BoardReactionType>(ADD_BOARD_REACTION, {detail: reaction});
  document.dispatchEvent(addBoardReactionEvent);
};

const remove = (id: string) => {
  const removeBoardReactionEvent = new CustomEvent<string>(REMOVE_BOARD_REACTION, {detail: id});
  document.dispatchEvent(removeBoardReactionEvent);
};

export const BoardReactionEmitter = {
  add,
  remove,
};
