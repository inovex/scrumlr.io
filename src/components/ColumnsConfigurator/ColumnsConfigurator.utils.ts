import {EditableTemplateColumn, ExistenceStatus, Template, TemplateColumn, TemplateWithColumns} from "store/features";
import {ColumnPlacement} from "./ColumnsConfigurator.types";

export const calcPlacement = (index: number, columns: TemplateColumn[]): ColumnPlacement => {
  const {length} = columns;
  if (length === 1) {
    return "all";
  }
  if (index === 0) {
    return "first";
  }
  if (index === length - 1) {
    return "last";
  }
  return "center";
};

// convert TemplateColumn to EditableColumn, used for test purposes
export const convertToEditableColumn = (c: TemplateColumn, overrideExistenceStatus?: Partial<ExistenceStatus>): EditableTemplateColumn =>
  ({
    ...c,
    persisted: false,
    mode: undefined,
    ...overrideExistenceStatus,
  }) as EditableTemplateColumn;

// from an array of all templates and column, create a TemplateAndColumns array with all corresponding pairs
export const mergeTemplateAndColumns = (templates: Template[], templateColumns: TemplateColumn[]) =>
  templates.map((template) => ({
    template,
    columns: templateColumns.filter((column) => column.template === template.id),
  })) as TemplateWithColumns[];

export const getTemplateAndColumnsByTemplateId = (templatesAndColumns: {templates: Template[]; templateColumns: TemplateColumn[]}, templateId: string) =>
  mergeTemplateAndColumns(templatesAndColumns.templates, templatesAndColumns.templateColumns).find((tac) => tac.template.id === templateId);
