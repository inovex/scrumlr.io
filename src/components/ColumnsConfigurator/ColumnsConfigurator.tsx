import {uniqueId} from "underscore";
import classNames from "classnames";
import {useState} from "react";
import {EditableTemplateColumn, TemplateColumn} from "store/features";
import {Color, getColorClassName, getNextColor, getPreviousColor} from "constants/colors";
import {closestCenter, DndContext, DragEndEvent, DragOverEvent, DragOverlay, DragStartEvent, PointerSensor, useSensor, useSensors} from "@dnd-kit/core";
import {horizontalListSortingStrategy, SortableContext} from "@dnd-kit/sortable";
import {ColumnsConfiguratorColumn} from "./ColumnsConfiguratorColumn/ColumnsConfiguratorColumn";
import {AddTemplateColumn} from "./AddTemplateColumn/AddTemplateColumn";
import {calcPlacement} from "./ColumnsConfigurator.utils";
import {AddTemplateColumnAlignment} from "./ColumnsConfigurator.types";
import "./ColumnsConfigurator.scss";

export type ColumnsConfiguratorProps = {
  className: string;
  templateId: string;
  columns: EditableTemplateColumn[]; // without deleted cols
  addColumn: (templateColumn: TemplateColumn, index: number) => void;
  moveColumn: (fromIndex: number, toIndex: number) => void;
  editColumn: (templateColumn: EditableTemplateColumn, overwrite: Partial<EditableTemplateColumn>) => void;
  deleteColumn: (templateColumn: EditableTemplateColumn) => void;
};

export const ColumnsConfigurator = (props: ColumnsConfiguratorProps) => {
  // id of column which is actively being dragged
  const [activeElementId, setActiveElementId] = useState<string | null>(null);
  const [activeColumn, setActiveColumn] = useState<EditableTemplateColumn | null>(null);
  const [dropElementId, setDropElementId] = useState<string | null>(null);

  const sensors = useSensors(useSensor(PointerSensor));

  const handleDragStart = (event: DragStartEvent) => {
    const activeId = event.active.id as string;
    setActiveElementId(activeId);
    setActiveColumn(props.columns.find((c) => c.id === activeId)!);
  };

  const handleDragOver = (event: DragOverEvent) => {
    setDropElementId((event.over?.id as string) ?? null);
  };

  const handleDragEnd = (event: DragEndEvent) => {
    const {active, over} = event;

    if (active.id !== over?.id) {
      const oldIndex = props.columns.findIndex((c) => c.id === active.id);
      const newIndex = props.columns.findIndex((c) => c.id === over?.id);

      props.moveColumn(oldIndex, newIndex);
    }

    // don't forget to clear states
    setActiveElementId(null);
    setActiveColumn(null);
    setDropElementId(null);
  };

  // function to calculate where each column would be positioned if the drag-and-drop operation were completed at its current state
  const getPotentialIndex = (index: number) => {
    // no valid drag operation occurring
    if (!activeElementId || !dropElementId) return index;

    const dragElementIndex = props.columns.findIndex((c) => c.id === activeElementId);
    const dropElementIndex = props.columns.findIndex((c) => c.id === dropElementId);

    // if current index matches active index of drag element, the column is being dragged there,
    // so the potential index is that of the drop element
    if (index === dragElementIndex) {
      return dropElementIndex;
    }

    // dragging right: active index less the drop zone, and element is between them, the element shifts to the left
    if (dragElementIndex < dropElementIndex && index > dragElementIndex && index <= dropElementIndex) {
      return index - 1;
    }

    // dragging left: analogue to above, element shifts to the right
    if (dragElementIndex > dropElementIndex && index < dragElementIndex && index >= dropElementIndex) {
      return index + 1;
    }

    // otherwise the index remains unchanged
    return index;
  };

  // renders the drag overlay, which is the element which is dragged
  const renderDragOverlay = () => {
    if (!activeElementId || !activeColumn) return null;
    const activeColumnIndex = getPotentialIndex(props.columns.findIndex((c) => c.id === activeColumn.id));

    return (
      <ColumnsConfiguratorColumn
        className={classNames("columns-configurator__column", "columns-configurator__column--ghost")}
        column={activeColumn}
        // here, use actual index unlike potential index in the main part
        index={activeColumnIndex}
        placement={calcPlacement(activeColumnIndex, props.columns)}
        allColumns={props.columns}
      />
    );
  };

  const addTemplateColumn = (alignment: AddTemplateColumnAlignment, color: Color) => {
    const newColumn: TemplateColumn = {
      id: uniqueId("col"),
      template: props.templateId,
      name: `Column ${props.columns.length + 1}`,
      description: "",
      color,
      visible: true,
      index: -1, // will be overwritten anyway
    };

    // target index for the right has to take deleted cols which are still part of the array into account,
    // which is why we use the total col amount as new index
    const index = alignment === "left" ? 0 : props.columns.length;

    props.addColumn(newColumn, index);
  };

  if (props.columns.length === 0) {
    return null;
  }

  return (
    <DndContext sensors={sensors} collisionDetection={closestCenter} onDragStart={handleDragStart} onDragOver={handleDragOver} onDragEnd={handleDragEnd}>
      <SortableContext items={props.columns} strategy={horizontalListSortingStrategy}>
        <div className={classNames(props.className, "columns-configurator", getColorClassName("backlog-blue"))}>
          <AddTemplateColumn alignment="left" color={getPreviousColor(props.columns[0].color)} onClick={addTemplateColumn} />
          <div className="columns-configurator__columns-wrapper">
            {props.columns.map((column, index) => {
              const potentialIndex = getPotentialIndex(index);
              return (
                <ColumnsConfiguratorColumn
                  className="columns-configurator__column"
                  key={column.id}
                  column={column}
                  index={potentialIndex}
                  activeDrag={column.id === activeElementId}
                  activeDrop={column.id === dropElementId}
                  placement={calcPlacement(potentialIndex, props.columns)}
                  allColumns={props.columns}
                  editColumn={props.editColumn}
                  deleteColumn={props.deleteColumn}
                />
              );
            })}
          </div>
          <AddTemplateColumn alignment="right" color={getNextColor(props.columns[props.columns.length - 1].color)} onClick={addTemplateColumn} />
        </div>
      </SortableContext>

      <DragOverlay>{renderDragOverlay()}</DragOverlay>
    </DndContext>
  );
};
