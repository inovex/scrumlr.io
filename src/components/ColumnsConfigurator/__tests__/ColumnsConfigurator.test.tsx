import {ColumnsConfigurator, ColumnsConfiguratorProps} from "../ColumnsConfigurator";
import {Partial} from "@react-spring/web";
import {render} from "testUtils";
import getTestApplicationState from "utils/test/getTestApplicationState";
import {convertToEditableColumn} from "components/ColumnsConfigurator/ColumnsConfigurator.utils";
import {fireEvent} from "@testing-library/react";

const renderColumnsConfigurator = (templateId: string, override?: Partial<Omit<ColumnsConfiguratorProps, "templateId" | "columns">>) => {
  const defaultProps: ColumnsConfiguratorProps = {
    className: "test-columns-configurator",
    templateId,
    columns: getTestApplicationState()
      .templateColumns.filter((tc) => tc.template === templateId)
      .map((tc) => convertToEditableColumn(tc)),
    addColumn: jest.fn(),
    deleteColumn: jest.fn(),
    editColumn: jest.fn(),
    moveColumn: jest.fn(),
  };

  return render(<ColumnsConfigurator {...defaultProps} {...override} />);
};

describe("ColumnsConfigurator", () => {
  it("should render correctly", () => {
    const {container} = renderColumnsConfigurator("test-templates-id-1");
    expect(container).toMatchSnapshot();

    const columnsNodeList = container.querySelectorAll(".columns-configurator__column");
    expect(columnsNodeList).toHaveLength(2);
  });

  it("should fire correct add new template col signal", () => {
    const addColumnSpy = jest.fn();
    const {container} = renderColumnsConfigurator("test-templates-id-1", {addColumn: addColumnSpy});
    const addTemplateColumnLeft = container.querySelector(".add-template-column--left")!;
    const addTemplateColumnRight = container.querySelector(".add-template-column--right")!;
    const addTemplateColumnLeftButton = addTemplateColumnLeft.querySelector("button")!;
    const addTemplateColumnRightButton = addTemplateColumnRight.querySelector("button")!;

    // template columns are backlog-blue and planning-pink, so the colors left and right to that are online-orange and poker-purple respectively
    expect(addTemplateColumnLeft).toHaveClass("accent-color__online-orange");
    expect(addTemplateColumnRight).toHaveClass("accent-color__poker-purple");

    fireEvent.click(addTemplateColumnLeftButton);
    expect(addColumnSpy).toHaveBeenCalledWith(
      {
        id: expect.stringMatching(/col(\d+)/),
        template: "test-templates-id-1",
        name: "Column 3",
        description: "",
        color: "online-orange",
        visible: true,
        index: -1,
      },
      0
    );

    fireEvent.click(addTemplateColumnRightButton);
    expect(addColumnSpy).toHaveBeenCalledWith(
      {
        id: expect.stringMatching(/col(\d+)/),
        template: "test-templates-id-1",
        name: "Column 3", // note hat since the state does not change, column length remains the same
        description: "",
        color: "poker-purple",
        visible: true,
        index: -1,
      },
      2
    );
  });
});
