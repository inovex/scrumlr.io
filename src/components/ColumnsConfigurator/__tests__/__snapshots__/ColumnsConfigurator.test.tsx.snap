// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnsConfigurator should render correctly 1`] = `
<div>
  <div
    class="test-columns-configurator columns-configurator accent-color__backlog-blue"
  >
    <div
      class="add-template-column add-template-column--left accent-color__online-orange"
    >
      <button
        class="add-template-column__button"
      >
        <svg
          class="add-template-column__icon"
        >
          add-column.svg
        </svg>
      </button>
    </div>
    <div
      class="columns-configurator__columns-wrapper"
    >
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled="false"
        aria-roledescription="sortable"
        class="columns-configurator__column columns-configurator-column columns-configurator-column--border-first columns-configurator-column--even accent-color__backlog-blue"
        role="button"
        style="--stripe-offset: 0px;"
        tabindex="0"
      >
        <div
          class="columns-configurator-column__name-details column-configurator-column-name-details__name-wrapper"
        >
          <input
            autocomplete="off"
            class="column-configurator-column-name-details__name"
            placeholder="Name"
            value="sample templates col name 1"
          />
          <div
            class="column-configurator-column-name-details__inline-description"
            role="button"
            tabindex="0"
          >
            sample templates col name 1
          </div>
        </div>
        <div
          class="columns-configurator-column__menu"
        >
          <svg
            class="columns-configurator-column__icon columns-configurator-column__icon--dnd columns-configurator-column__drag-element"
            data-cy="columns-configurator-column__drag-element"
          >
            drag-and-drop.svg
          </svg>
          <span
            aria-label="current colorBacklog Blue"
            class="columns-configurator-column__color-picker columns-configurator-column__color-picker--column-first color-picker__color-option color-picker__color-option--selected"
            data-cy="columns-configurator-column__color-picker"
            role="button"
            tabindex="0"
          />
          <button
            class="columns-configurator-column__button column-configurator-column__visibility-button"
          >
            <svg
              class="columns-configurator-column__icon columns-configurator-column__icon--visible"
              data-cy="columns-configurator-column__icon--visibility"
            >
              visible.svg
            </svg>
          </button>
          <button
            class="columns-configurator-column__button column-configurator-column__delete-button"
          >
            <svg
              class="columns-configurator-column__icon columns-configurator-column__icon--delete"
            >
              trash.svg
            </svg>
          </button>
        </div>
      </div>
      <div
        aria-describedby="DndDescribedBy-0"
        aria-disabled="false"
        aria-roledescription="sortable"
        class="columns-configurator__column columns-configurator-column columns-configurator-column--border-last columns-configurator-column--odd columns-configurator-column--hidden accent-color__planning-pink"
        role="button"
        style="--stripe-offset: 0px;"
        tabindex="0"
      >
        <div
          class="columns-configurator-column__name-details column-configurator-column-name-details__name-wrapper"
        >
          <input
            autocomplete="off"
            class="column-configurator-column-name-details__name"
            placeholder="Name"
            value="sample templates col name 2"
          />
          <div
            class="column-configurator-column-name-details__inline-description"
            role="button"
            tabindex="0"
          >
            sample templates col name 2
          </div>
        </div>
        <div
          class="columns-configurator-column__menu"
        >
          <svg
            class="columns-configurator-column__icon columns-configurator-column__icon--dnd columns-configurator-column__drag-element"
            data-cy="columns-configurator-column__drag-element"
          >
            drag-and-drop.svg
          </svg>
          <span
            aria-label="current colorPlanning Pink"
            class="columns-configurator-column__color-picker columns-configurator-column__color-picker--column-last color-picker__color-option color-picker__color-option--selected"
            data-cy="columns-configurator-column__color-picker"
            role="button"
            tabindex="0"
          />
          <button
            class="columns-configurator-column__button column-configurator-column__visibility-button"
          >
            <svg
              class="columns-configurator-column__icon columns-configurator-column__icon--hidden"
              data-cy="columns-configurator-column__icon--visibility"
            >
              hidden.svg
            </svg>
          </button>
          <button
            class="columns-configurator-column__button column-configurator-column__delete-button"
          >
            <svg
              class="columns-configurator-column__icon columns-configurator-column__icon--delete"
            >
              trash.svg
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="add-template-column add-template-column--right accent-color__poker-purple"
    >
      <button
        class="add-template-column__button"
      >
        <svg
          class="add-template-column__icon"
        >
          add-column.svg
        </svg>
      </button>
    </div>
  </div>
  <div
    id="DndDescribedBy-0"
    style="display: none;"
  >
    
    To pick up a draggable item, press the space bar.
    While dragging, use the arrow keys to move the item.
    Press space again to drop the item in its new position, or press escape to cancel.
  
  </div>
  <div
    aria-atomic="true"
    aria-live="assertive"
    id="DndLiveRegion-0"
    role="status"
    style="position: fixed; top: 0px; left: 0px; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%); white-space: nowrap;"
  />
</div>
`;
